caip-122: Sign in With X (SIWx) (#122)
ChainAgnostic/CAIPs: CAIPs/caip-122.md

| `domain`          | string          | ✓         | [RFC 4501][rfc 4501] `dnsauthority` that is requesting the signing.                                                                                                                               |
| `address`         | string          | ✓         | Blockchain address, as defined by [CAIP-10][caip-10], performing the signing; should include [CAIP-2][caip-2] chain id namespace                                                                  |
| `uri`             | string          | ✓         | [RFC 3986][rfc 3986] URI referring to the resource that is the subject of the signing i.e. the subject of the claim.                                                                              |
| `version`         | string          | ✓         | Current version of the message.                                                                                                                                                                   |

> | `statement`       | string          |           | Human-readable ASCII assertion that the user will sign. It _must not_ contain `\n`.                                                                                                               |

Reviewer: ```suggestion
| `statement`       | string          |           | Human-readable ASCII assertion that the user will sign. It _MUST NOT_ contain `\n`.                                                                                                               |
```

| `request-id`      | string          |           | System-specific identifier used to uniquely refer to the authentication request.                                                                                                                  |
| `resources`       | List of strings |           | List of information or references to information the user wishes to have resolved as part of the authentication by the relying party; express as [RFC 3986][rfc 3986] URIs and separated by `\n`. |
| `signature`       | bytes           | ✓         | Signature of the message signed by the wallet.                                                                                                                                                    |
| `type`            | string          | ✓         | Type of the signature to be generated, as defined in the namespaces for this CAIP.                                                                                                                |

### Namespace Specification

> A namespace specification _must_ provide:

Reviewer: ```suggestion
A namespace specification _MUST_ provide:
```

1. signing algorithm, or multitude of those,
2. accompanied by `type` strings that designate each signing algorithm,
3. a procedure for creating a signing input from the data model specified in this document.

> The signing algorithm _must_ cover:

Reviewer: ```suggestion
The signing algorithm _MUST_ cover:
```

Below is an example of the data model represented as a plain text similar Ethereum, and then converted to raw bytes.

Plain text representation:

```
service.org wants you to sign in with your Solana account:
GwAF45zjfyGzUbd3i3hXxzGeuchzEZXwpRYHZM5912F1

I accept the ServiceOrg Terms of Service: https://service.org/tos

URI: https://service.org/login
Version: 1
Nonce: 32891757
Issued At: 2021-09-30T16:25:24.000Z
Chain ID: 1
Resources:
- ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
- https://example.com/my-web2-claim.json
```

Raw bytes (encoded as base64url for brevity).

```
c2VydmljZS5vcmcgd2FudHMgeW91IHRvIHNpZ24gaW4gd2l0aCB5b3VyIFNvbGFuYSBhY2NvdW50OgpHd0FGNDV6amZ5R3pVYmQzaTNoWHh6R2V1Y2h6RVpYd3BSWUhaTTU5MTJGMQoKSSBhY2NlcHQgdGhlIFNlcnZpY2VPcmcgVGVybXMgb2YgU2VydmljZTogaHR0cHM6Ly9zZXJ2aWNlLm9yZy90b3MKClVSSTogaHR0cHM6Ly9zZXJ2aWNlLm9yZy9sb2dpbgpWZXJzaW9uOiAxCk5vbmNlOiAzMjg5MTc1NwpJc3N1ZWQgQXQ6IDIwMjEtMDktMzBUMTY6MjU6MjQuMDAwWgpDaGFpbiBJRDogMQpSZXNvdXJjZXM6Ci0gaXBmczovL1FtZTdzczNBUlZneHY2clhxVlBpaWtNSjh1Mk5MZ21nc3pnMTNwWXJES0VvaXUKLSBodHRwczovL2V4YW1wbGUuY29tL215LXdlYjItY2xhaW0uanNvbg
```

## Rationale

> - As a chain-agnostic standard, SIWx should allow for blockchain wallet based authentication across non-blockchain applications regardless of which chain/wallet the user is using.

Reviewer: ```suggestion
- As a chain-agnostic standard, SIWx should allow for authentication via blockchain wallet across non-blockchain applications regardless of which chain/wallet the user is using.
```

> - The application server must be able to implement fully usable support for its end without forcing a change in the wallets

Reviewer: ```suggestion
- The application server must be able to implement fully functional authentication for as many users as possible without forcing a change to wallets
```

> - The application server must be able to implement fully functional authentication for as many users as possible without forcing a change to wallets

Reviewer: ```suggestion
- The application server _MUST_ be able to implement fully functional authentication for as many users as possible without forcing a change to wallets
```
